<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <script>
        class Promise{
            constructor(executor){
                this.state = 'pending'
                this.value = undefined //记录成功的值 给.then使用
                this.reason = undefined //记录失败的值 给.catch使用

                this.onResolveCallbacks = [] //存放成功时需要执行的函数
                this.onRejectedCallbacks = []

                let resolve = (value) => {
                    if(this.state === 'pending'){
                        this.state = 'fullfilled'
                        this.value = value
                        this.onResolveCallbacks.forEach(fn=>fn()) 
                    }
                }
                let reject = (reason) => {
                    if(this.state === 'pending'){
                        this.state = 'rejected'
                        this.reason = reason
                        this.onRejectedCallbacks.forEach(fn=>fn())
                    }
                }
                try{
                    executor(resolve,reject)
                }catch(err){
                    reject()
                }
            }
            then(onFullFilled,onRejected){
                let promise2 = new Promise((resolve,reject)=>{
                    if(this.state === 'fullfilled'){
                        // 增加异步处理
                        setTimeout(()=>{
                            let x = onFullFilled(this.value) //p.then中return的是x
                            resolvePromise(x,resolve,reject,promise2)
                        },0)

                    }
                    if(this.state === 'rejected'){
                        onRejected(this.reason)
                    }
                    // 异步处理
                    if(this.state === 'pending'){
                        this.onResolveCallbacks.push(()=>{
                            onFullFilled(this.value)
                        })
                        this.onRejectedCallbacks.push(()=>{
                            onRejected(this.reason)
                        })
                    }
                })
                return promise2
            }
        }
        function resolvePromise(x,resolve,reject,promise2){
            // 处理循环调用提示报错
            // console.log( '111' )
            if(x===promise2){
                let err = 'Uncaught (in promise) TypeError: Chaining cycle detected for promise #<Promise>'
                console.error(err)
                return reject(err)
            }
            if (x instanceof Promise){
                x.then(resolve,reject)  
            }else{
                resolve(x)
            }
        }
        const p = new Promise((resolve,reject)=>{
           resolve(200)
        })
        
        const p2 = p.then(data=>{
            console.log(data)
            return p2 //需要报错提示
        })
    </script>
</body>
</html>
</body>
</html>